platforms/solana/protocols/tbtc/src/bridge.ts

```diff
@@ -0,0 +1,346 @@
+import { BN, Program } from '@coral-xyz/anchor';
+import { WormholeGateway, WormholeGatewayIdl } from './anchor-idl/gateway.js';
+import {
+  getConfigPda,
+  getCoreMessagePda,
+  getCustodianPda,
+  getGatewayInfoPda,
+  getMinterInfoPda,
+  TBTC_PROGRAM_ID,
+} from './utils.js';
+import {
+  createAssociatedTokenAccountInstruction,
+  getAssociatedTokenAddress,
+} from '@solana/spl-token';
+
+export class SolanaTBTCBridge<N extends Network, C extends SolanaChains>
+  implements TBTCBridge<N, C>
+{
+  gateway: Program<WormholeGateway>;
+  tokenBridgeId: PublicKey;
+  coreBridgeId: PublicKey;
+
+  constructor(
+    readonly network: N,
+    readonly chain: C,
+    readonly connection: Connection,
+    readonly contracts: Contracts,
+  ) {
+    if (this.network !== 'Mainnet') {
+      throw new Error('TBTC is only supported on Mainnet');
+    }
+
+    if (!this.contracts.tbtc) {
+      throw new Error('TBTC contract address is required');
+    }
+
+    if (!this.contracts.tokenBridge) {
+      throw new Error('TokenBridge contract address is required');
+    }
+
+    if (!this.contracts.coreBridge) {
+      throw new Error('CoreBridge contract address is required');
+    }
+
+    this.gateway = new Program<WormholeGateway>(
+      WormholeGatewayIdl,
+      this.contracts.tbtc,
+      {
+        connection,
+      },
+    );
+    this.tokenBridgeId = new PublicKey(this.contracts.tokenBridge);
+    this.coreBridgeId = new PublicKey(this.contracts.coreBridge);
+  }
+
+  static async fromRpc<N extends Network>(
+    connection: Connection,
+    config: ChainsConfig<N, Platform>,
+  ): Promise<SolanaTBTCBridge<N, SolanaChains>> {
+    const [network, chain] = await SolanaPlatform.chainFromRpc(connection);
+
+    const conf = config[chain]!;
+    if (conf.network !== network)
+      throw new Error(`Network mismatch: ${conf.network} != ${network}`);
+
+    return new SolanaTBTCBridge(
+      network as N,
+      chain,
+      connection,
+      conf.contracts,
+    );
+  }
+
+  async *transfer(
+    sender: AccountAddress<C>,
+    recipient: ChainAddress,
+    amount: bigint,
+  ): AsyncGenerator<SolanaUnsignedTransaction<N, C>> {
+    const senderPk = new SolanaAddress(sender).unwrap();
+
+    const custodian = getCustodianPda(this.gateway.programId);
+
+    const { tbtcMint, wrappedTbtcToken, wrappedTbtcMint } =
+      await this.gateway.account.custodian.fetch(custodian);
+
+    const tokenBridgeWrappedAsset = deriveWrappedMetaKey(
+      this.tokenBridgeId,
+      wrappedTbtcMint,
+    );
+
+    const tokenBridgeConfig = deriveTokenBridgeConfigKey(this.tokenBridgeId);
+
+    const tokenBridgeTransferAuthority = deriveAuthoritySignerKey(
+      this.tokenBridgeId,
+    );
+
+    const coreFeeCollector = coreUtils.deriveFeeCollectorKey(this.coreBridgeId);
+
+    // NOTE: There is a race condition where the sequence changes
+    // while the transaction is in flight. This would cause the
+    // transaction to fail and the user would have to retry.
```

Comment:
I wonder why they have this global sequence number like this... to prevent replay attacks?

Reply:
I'm not sure tbh, but I remember running into this issue in legacy Connect. The issue was raised to the TBTC team, but nothing has been done to fix it.